File: 2023 yard hoof cup/C++ The end scene/5.cpp
#include <bits/stdc++.h>

using namespace std;
#define int long long

int n;
vector<int> tree;
int ms;
int res;

int getLeft(int i) {
    return 2 * i + 1;
}

int getRight(int i) {
    return 2 * i + 2;
}

void dfs(int i,int deep) {
    if (i >= n) {
        return;
    }
    ms = max(ms, tree[i] * deep);
    dfs(getLeft(i), deep + 1);
    dfs(getRight(i), deep + 1);
}

void dfs(int i) {
    ms = 0;
    dfs(i, 0);
    res += ms;
}

void solve() {
    cin >> n;
    tree.assign(n, 0);
    for (int i = 0; i < n; i++) {
        cin >> tree[i];
    }
    res = 0;
    for (int i = 0; i < n; i++) {
        dfs(i);
    }
    cout << res << endl;
}

auto main() -> signed {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);

    solve();
    return 0;
}

File: 2023 yard hoof cup/Python The end scene/5.py
def getLeft(i: int) -> int:
    return i * 2 + 1


def getRight(i: int) -> int:
    return i * 2 + 2


def dfs(i: int, deep: int):
    global ms
    if i >= n:
        return
    ms = max(ms, tree[i] * deep)
    dfs(getLeft(i), deep + 1)
    dfs(getRight(i), deep + 1)


def find(i: int):
    global res, ms
    ms = 0
    dfs(i, 0)
    res += ms


def solve():
    global n, tree, ms, res

    ms = 0
    res = 0
    n = int(input())
    tree = list(map(int, input().split()))
    for i in range(n):
        find(i)
    print(res)


if __name__ == '__main__':
    solve()
